python/modem-firmware-1.3+/gather_attestation_tokens.py

```diff
@@ -0,0 +1,469 @@
+                                        quoting=csv.QUOTE_MINIMAL)
+                csv_writer.writerows(keep_rows)
+                csvfile.close()
+        except OSError:
+            print(error_style(f'Error opening file (write) {csv_filename}'))
+
+    return duplicate_rows, row_count
+
+def user_request_open_mode(filename, append):
+    mode = 'a' if append else 'w'
+    exists = os.path.isfile(filename)
+
+    # if not appending, give user a choice whether to overwrite
+    if not append and exists:
+        answer = ' '
+        while answer not in 'yan':
+            answer = input('--- File {} exists; overwrite, append, or quit (y,a,n)? '.format(filename))
+
+        if answer == 'n':
+            print(local_style('File will not be overwritten'))
+            return None
+        elif answer == 'y':
+            mode = 'w'
+        else:
+            mode = 'a'
+
+    elif not exists and append:
+        mode = 'w'
+        print('Append specified but file does not exist...')
+
+    return mode
+
+def save_attestation_csv(csv_filename, append, replace, imei, uuid, attestation_token):
+    mode = user_request_open_mode(csv_filename, append)
+
+    if mode == None:
+        return
+
+    row_count = 0
+
+    row = f'{imei},{uuid},{attestation_token},{datetime.now(timezone.utc).isoformat()}\n'
+
+    if mode == 'a':
+        duplicate_rows, row_count = check_if_device_exists_in_csv(csv_filename, uuid, replace)
+
+        if len(duplicate_rows):
+            if replace:
+                print(hivis_style(f'Removed existing data:\r\n\t{duplicate_rows}'))
+            else:
+                print(error_style('Device already exists in CSV, the following row was NOT added:'))
+                print(local_style(row))
+                return
+
+    try:
+        with open(csv_filename, mode, newline='\n') as devinfo_file:
+            devinfo_file.write(row)
+        print(local_style(f'Attestation CSV file {csv_filename} saved, row count: {row_count + 1}'))
+    except OSError:
+        print(error_style('Error opening file {}'.format(csv_filename)))
+
+def error_exit(err_msg):
+    cleanup()
+    if err_msg:
+        sys.stderr.write(error_style(err_msg))
+        sys.stderr.write('\n')
+        sys.exit(1)
+    else:
+        sys.exit('Error... exiting.')
+
+def main():
+    global args
+    global plain
+    global ser
+    global rtt
+    global cmd_term_key
+    global at_cmd_prefix
+
+    # initialize arguments
+    args = parse_args()
+    plain = args.plain
+
+    rtt = None
+
+    # Set to CRLF for interaction with provisioning sample
+    cmd_term_key = 'CRLF'
+
+    # Adjust method to send an AT command
+    at_cmd_prefix = '' if not args.shell else 'at '
+
+    # initialize colorama
+    if is_windows:
+        init(convert = not plain)
+
+    if args.verbose:
+        print(send_style('OS detect: Linux={}, MacOS={}, Windows={}\n'.
+                          format(is_linux, is_macos, is_windows)))
```

Comment:
For newly introduced code, maybe we should be using the `CredentialCommandInterface` classes? Less potential refactor work for later. It also significantly reduces code duplication

Reply:
Not sure what you mean. Which newly introduced code?

Reply:
If you mean to allow for getting attestation tokens from a Wi-Fi device using the TLS credentials shell, I am not sure that would ever happen. Maybe I'm wrong.

Reply:
> Which newly introduced code?
`gather_attestation_tokens.py`, but maybe I am misunderstanding the github delta.

> If you mean to allow for getting attestation tokens from a Wi-Fi device

I was thinking, more, add an attestation token command to the base class, then have the `TLS credentials shell` subclass not support it, but have the `AT` subclass support it, and then just use the AT subclass in this file.

But now that you mention it, I see your point that attestation token gathering is not ***really*** necessarily a "credential command".

Upon reconsideration, this does seem like a gray area to me, so I'm fine if you just want to leave it this way for now until we eventually work on some kind of broader refactor for this codebase

Reply:
Feel free to add to this ticket I added in response to Justin: https://nordicsemi.atlassian.net/browse/IRIS-9710.